### PR TITLE
add container id to ApiServer.State and send in header

### PR DESCRIPTION
**Pull request by dallincrane**
Adds a `Datadog-Container-ID` header when a container is being used.
Modeled on the [DD Node.js client code](https://github.com/DataDog/dd-trace-js)

Happy to add a couple tests if required but would need some guidance — the file read might require a mock setup.

### CODE REVIEW

File: lib/spandex_datadog/api_server.ex [hunk 1]
```diff
@@ -133,8 +150,9 @@ defmodule SpandexDatadog.ApiServer do
   end
 
   @spec send_and_log([Trace.t()], State.t()) :: :ok
-  def send_and_log(traces, %{verbose?: verbose?} = state) do
+  def send_and_log(traces, %{container_id: container_id, verbose?: verbose?} = state) do
     headers = @headers ++ [{"X-Datadog-Trace-Count", length(traces)}]
+    headers = headers ++ List.wrap(if container_id, do: {"Datadog-Container-ID", container_id})
```
Comment (dallincrane) [inline lib/spandex_datadog/api_server.ex:155]: reference to this header in DataDog's Node.js client
https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/exporters/agent/request.js#L26

File: lib/spandex_datadog/api_server.ex [hunk 2]
```diff
@@ -95,12 +96,28 @@ defmodule SpandexDatadog.ApiServer do
...
+      container_id: get_container_id()
     }
 
     {:ok, state}
   end
 
+  @cgroup_uuid "[0-9a-f]{8}[-_][0-9a-f]{4}[-_][0-9a-f]{4}[-_][0-9a-f]{4}[-_][0-9a-f]{12}"
+  @cgroup_ctnr "[0-9a-f]{64}"
+  @cgroup_task "[0-9a-f]{32}-\\d+"
+  @cgroup_regex Regex.compile!("/(#{@cgroup_uuid}|#{@cgroup_ctnr}|#{@cgroup_task})(?:.scope)?$", "m")
```
Comment (dallincrane) [inline lib/spandex_datadog/api_server.ex:109]: reference to this regex used to get the container_id
https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/exporters/agent/docker.js

**Comment by rocpatel** [comment]
adding a link to an issue that references this concern.  https://github.com/spandex-project/spandex_datadog/issues/25

File: lib/spandex_datadog/api_server.ex [hunk 3]
```diff
@@ -95,12 +96,28 @@ defmodule SpandexDatadog.ApiServer do
...
+
+  defp get_container_id() do
+    with {:ok, file_binary} <- File.read("/proc/self/cgroup"),
+         true <- String.valid?(file_binary),
```
Comment (kamilkowalski) [inline lib/spandex_datadog/api_server.ex:113]: I'm not sure this is needed - `Regex` works with binaries, and the binary either matches the pattern or not. Doesn't really matter if the string contains valid characters as far as I understand.

**Review by kamilkowalski — COMMENTED** [review]
Hi @dallincrane - thanks for contributing! We could actually use this feature at my company, so I took the liberty to check it. I have just one comment as to the implementation.

When it comes to testing, I have two ideas:
 
1. Inject the `cgroup` file path either through opts, or through application env, and default to `/proc/self/cgroup`. Then override that value in tests as e.g. `test/support/cgroup`.
2. Extract `cgroup` parsing to a separate module and expose a function like `parse/1` which takes the file path as an argument. Test that module separately, then use it in `ApiServer`. You could even use `mox` to mock that parsing module in tests.

Let me know what you think.

**Comment by kamilkowalski** [comment]
I can also help to test this - I'll spin up an environment with this change and see if it works.

**Comment by kamilkowalski** [comment]
Seems to be working :clap: 
Screenshot_2021-04-14 refresh GET probe (env pr-23974-ns) Datadog

**Comment by dallincrane** [comment]
> Seems to be working 👏
> Screenshot_2021-04-14 refresh GET probe (env pr-23974-ns) Datadog

Nice! At my last company we had tested it and had it in production. Glad you are seeing the same!

**Comment by dallincrane** [comment]
Hi @kamilkowalski - thanks for input! Do you by chance have write access? If you don't, it looks like we need someone with write access to provide a review and perhaps give a little input into some of your suggestions.

For some background, I was trying to choose code that provided the least resistance and least amount of project structure change so it could be accepted and quickly released.

>     1. Inject the `cgroup` file path either through opts, or through application env, and default to `/proc/self/cgroup`. Then override that value in tests as e.g. `test/support/cgroup`.

I used [Datadog's Javascript code](https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/exporters/agent/docker.js) as a refernence. It is hard coded there. I'm not sure how many projects would have a custom cgroup file path
 
>     2. Extract `cgroup` parsing to a separate module and expose a function like `parse/1` which takes the file path as an argument. Test that module separately, then use it in `ApiServer`. You could even use `mox` to mock that parsing module in tests.

Generally I do really like what you are suggesting and it even matches the Datadog code! But because adding a new file and module requires a naming discussion and `mox` isn't currently in the project, I decided to keep the PR simple.

**Comment by kamilkowalski** [comment]
@dallincrane I'm not a maintainer so we probably need to wait for @zachdaniel or @GregMefford to chip in.

As to testing - I think sometimes it makes sense to extend the library's public interface to allow for testing (if the alternative is no tests). Also, the second solution doesn't necessarily require Mox - it only requires us to extract the cgroups functionality to a separate module.

**Comment by moleskin-smile** [comment]
FYI: regex used in the reference implementation has been changed since opening this PR: https://github.com/DataDog/dd-trace-js/pull/1314.

**Comment by dallincrane** [comment]
thank you @moleskin-smile! updated

**Pull request merged**